### PR TITLE
feat: Geat4 detector plugin, part 3 (final)

### PR DESCRIPTION
**Pull request by asalzburger**
This PR concludes (for the moment) the Geant4 based detector example, it adds at the `Detector/Geant4` to the Examples, adds the python bindings.

It is currently blocked as it depends on #1758  and #1759 - the first two parts of this development.

With this PR, relatively simple GDML files can be translated one-to-one into a `TrackingGeometry`, it serves as a validation test bench, and will also provide future translation into `Experimental::Detector` in order to cross-validate.

The following shows an output of a 'light' version of the ODD tracker:

Screenshot 2022-12-16 at 17 04 33

Screenshot 2022-12-16 at 17 05 01

This can be reproduced with the following sources:

[odd-light.zip](https://github.com/acts-project/acts/files/10247296/odd-light.zip)

This contains:
```sh
-rwxr-xr-x@ 1 salzburg  staff      991 Dec 16 16:58 gdml_geometry.py
-rwxr-xr-x@ 1 salzburg  staff     2156 Dec 16 16:57 gdml_material_recording.py
-rwxr-xr-x@ 1 salzburg  staff     1926 Dec 16 16:57 gdml_material_validation.py
-rw-r--r--@ 1 salzburg  staff  6434606 Dec 16 16:56 odd-light.gdml
-rw-r--r--  1 salzburg  staff    79113 Dec 16 16:55 odd-proto-detector.json
```

The `odd-light.gdml` is the GDML description of the detector, the `odd-proto-detector.json` the structure and then are three run files.

### CODE REVIEW

**Comment by asalzburger** [comment]
Ping @benjaminhuth - I think for a full usage, there needs to be a follow-up PR:
- add the `GeometryIdentifierHook` correctly to the `KDTreeTrackingGeometryBuilder`
- adapt the smearing/digization information
-

**Comment by andiwand** [comment]
lets get part 1 and 2 in so this one shrinks a little

**Comment by asalzburger** [comment]
@andiwand - part3 is ready to be reviewed.

File: Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Geant4Detector/Geant4Detector.hpp"
+
+#include "Acts/Geometry/Detector.hpp"
+#include "ActsExamples/Framework/IContextDecorator.hpp"
+
+#include <boost/program_options.hpp>
```
Comment (benjaminhuth) [inline Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp:14]: Is boost program options needed here? I don't see it used anywhere...

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+#include <string>
+#include <vector>
+
+#include "G4VPhysicalVolume.hh"
+
+namespace Acts {
+
+class IGeant4PhysicalVolumeSelector {
+ public:
+  virtual ~IGeant4PhysicalVolumeSelector(){};
```
Comment (benjaminhuth) [inline Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp:22]: Is it on purpose that the Base class does not have the `bool select(...) const` interface?

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class Detector;
+}  // namespace Experimental
+}  // namespace Acts
+
+namespace ActsExamples {
+namespace Geant4 {
+
+/// @class Geant4DetectorService
+///
+class Geant4DetectorService final : public BareService {
```
Comment (benjaminhuth) [inline Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:29]: Actually, I'm not really sure why we inherit from `BareService` here... We do not plug it into the sequencer, right?
As far as I see the code would work the same without inheritance (only used locally in the `constructDetector`/`constructTrackingGeometry` functions), but maybe I overlook something.

Could we even make this a local class in the the `src` folder and only expose the config object?

**Review by benjaminhuth — COMMENTED** [review]
Only a few comments, feel free to ignore them!

**Comment by andiwand** [comment]
taking a look now 👀

File: Examples/Detectors/Geant4Detector/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+target_include_directories(
+    ActsExamplesDetectorGeant4
+  PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
+target_link_libraries(
+    ActsExamplesDetectorGeant4
+  PUBLIC ActsCore ActsExamplesFramework ActsExamplesGeant4 ActsPluginGeant4)
+
+install(
+  TARGETS ActsExamplesDetectorGeant4
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
```
Comment (andiwand) [inline Examples/Detectors/Geant4Detector/CMakeLists.txt:14]: nitpick: empty line at the end of the file

File: Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+auto ActsExamples::Geant4::Geant4Detector::constructTrackingGeometry(
+    const ActsExamples::Geant4::Geant4DetectorService::Config& cfg)
+    -> std::pair<TrackingGeometryPtr, ContextDecorators> {
+  auto g4DetectorService = ActsExamples::Geant4::Geant4DetectorService(cfg);
+
+  g4DetectorService.startRun();
+
+  return std::make_pair<TrackingGeometryPtr, ContextDecorators>(
+      g4DetectorService.trackingGeometry(), {});
+}
```
Comment (andiwand) [inline Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp:35]: nitpick: new line at the end of the file

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    /// Name for passive selection
+    std::vector<std::string> passiveSelections = {"pass_vol"};
+    /// Logging level of the master tool
+    Acts::Logging::Level logLevel = Acts::Logging::INFO;
+    /// Logging level of the child tools
+    Acts::Logging::Level toolLogLevel = Acts::Logging::INFO;
+  };
+
+  Geant4DetectorService(const Config& cfg);
+  ~Geant4DetectorService() final = default;
```
Comment (andiwand) [inline Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:56]: nitpick: could be omitted

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+inline std::shared_ptr<Acts::Experimental::Detector>
+Geant4DetectorService::detector() const {
+  return m_detector;
+}
+
+inline std::shared_ptr<const Acts::TrackingGeometry>
+Geant4DetectorService::trackingGeometry() const {
+  return m_trackingGeometry;
+}
```
Comment (andiwand) [inline Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:87]: should we move the implementation into the class to be more concise?

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4Converters.hpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  /// order to keep the system right-handed
+  bool keepAxisOrder = false;
+
+  /// @brief Convert to cylinder bounds
+  ///
+  /// @param g4Tubs a Geant4 tube shape
+  ///
+  /// @return an Acts Cylinder bounds object
+  std::tuple<std::shared_ptr<CylinderBounds>, ActsScalar> cylinderBounds(
+      const G4Tubs& g4Tubs);
```
Comment (andiwand) [inline Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4Converters.hpp:86]: what does the scalar mean in the tuple?

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4DetectorElement.hpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+inline ActsScalar Geant4DetectorElement::thickness() const {
+  return m_thickness;
+}
+
+inline const G4VPhysicalVolume& Geant4DetectorElement::g4PhysicalVolume()
+    const {
+  return (*m_g4physVol);
+}
```
Comment (andiwand) [inline Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4DetectorElement.hpp:76]: would it make sense to move the impl into the class definition?

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp (hunk 2 above)
Comment (andiwand) [inline Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp:22]:
```suggestion
  virtual ~IGeant4PhysicalVolumeSelector() = default;
```

File: Plugins/Geant4/src/Geant4Converters.cpp [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+  RotationMatrix3 rotation;
+  rotation << g4Trf.xx(), g4Trf.yx(), g4Trf.zx(), g4Trf.xy(), g4Trf.yy(),
+      g4Trf.zy(), g4Trf.xz(), g4Trf.yz(), g4Trf.zz();
+  Transform3 transform = Transform3::Identity();
+  transform.matrix().block(0, 0, 3, 1) = rotation.col(0);
+  transform.matrix().block(0, 1, 3, 1) = rotation.col(1);
+  transform.matrix().block(0, 2, 3, 1) = rotation.col(2);
+  transform.matrix().block(0, 3, 3, 1) = translation;
+  return transform;
+}
```
Comment (andiwand) [inline Plugins/Geant4/src/Geant4Converters.cpp:75]: could this be implemented using `Acts::Transform3 Acts::Geant4AlgebraConverter::transform(const G4RotationMatrix& g4Rot, const G4ThreeVector& g4Trans)` or vice versa?

File: Plugins/Geant4/src/Geant4DetectorSurfaceFactory.cpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+             surface});
+      } else {
+        cache.passiveSurfaces.push_back(surface);
+      }
+      // Count the material conversion
+      if (surface->surfaceMaterial() != nullptr) {
+        ++cache.convertedMaterials;
+      }
+    }
+    return;
```
Comment (andiwand) [inline Plugins/Geant4/src/Geant4DetectorSurfaceFactory.cpp:68]: or do you intend to extend the function?

```suggestion
```

File: Tests/UnitTests/Plugins/Geant4/Geant4DetectorElementTests.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  Acts::Geant4DetectorElement g4DetElement(rSurface, 0.1, *g4physVol.get());
+
+  BOOST_CHECK(g4DetElement.thickness() == 0.1);
+  BOOST_CHECK(&g4DetElement.surface() == rSurface.get());
+  BOOST_CHECK(&g4DetElement.g4PhysicalVolume() == g4physVol.get());
+  BOOST_CHECK(
+      g4DetElement.transform(tContext).isApprox(rSurface->transform(tContext)));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (andiwand) [inline Tests/UnitTests/Plugins/Geant4/Geant4DetectorElementTests.cpp:44]: nitpick: new line at the end of the file

ps I checked clang-format but it seems not to support it

File: Tests/UnitTests/Plugins/Geant4/Geant4PhysicalVolumeSelectorsTests.cpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  BOOST_CHECK(!nameSelector(*worldPV));
+  BOOST_CHECK(nameSelector(*boxPV));
+
+  auto nameSelectorE =
+      Acts::Geant4PhysicalVolumeSelectors::generateNameSelector({"ox"}, true);
+  BOOST_CHECK(!nameSelectorE(*worldPV));
+  BOOST_CHECK(!nameSelectorE(*boxPV));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (andiwand) [inline Tests/UnitTests/Plugins/Geant4/Geant4PhysicalVolumeSelectorsTests.cpp:50]: nitpick: newline at the end of the file

File: Examples/Detectors/Geant4Detector/CMakeLists.txt (hunk 4 above)
Comment (asalzburger) [reply Examples/Detectors/Geant4Detector/CMakeLists.txt:14]: And fixed.

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp (hunk 6 above)
Comment (asalzburger) [reply Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:56]: Fixed.

File: Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp (hunk 5 above)
Comment (asalzburger) [reply Examples/Detectors/Geant4Detector/src/Geant4Detector.cpp:35]: Inserted.

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp (hunk 7 above)
Comment (asalzburger) [reply Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:87]: Done.

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4Converters.hpp (hunk 8 above)
Comment (asalzburger) [reply Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4Converters.hpp:86]: Updated the documentation, the scalar is the original thickness, one needs to know that to do the material compression.

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4DetectorElement.hpp (hunk 9 above)
Comment (asalzburger) [reply Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4DetectorElement.hpp:76]: Done.

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp (hunk 2 above)
Comment (asalzburger) [reply Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp:22]: Done.

File: Plugins/Geant4/src/Geant4Converters.cpp (hunk 10 above)
Comment (asalzburger) [reply Plugins/Geant4/src/Geant4Converters.cpp:75]: Indeed it can.

File: Plugins/Geant4/src/Geant4DetectorSurfaceFactory.cpp (hunk 11 above)
Comment (asalzburger) [reply Plugins/Geant4/src/Geant4DetectorSurfaceFactory.cpp:68]: Yes, there will be more to come, but the `return` is not necessary.

File: Tests/UnitTests/Plugins/Geant4/Geant4DetectorElementTests.cpp (hunk 12 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Geant4/Geant4DetectorElementTests.cpp:44]: Fixed.

File: Tests/UnitTests/Plugins/Geant4/Geant4PhysicalVolumeSelectorsTests.cpp (hunk 13 above)
Comment (asalzburger) [reply Tests/UnitTests/Plugins/Geant4/Geant4PhysicalVolumeSelectorsTests.cpp:50]: Fixed.

File: Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp (hunk 2 above)
Comment (asalzburger) [reply Plugins/Geant4/include/Acts/Plugins/Geant4/Geant4PhysicalVolumeSelectors.hpp]: I just tried to keep it consistent.

File: Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp (hunk 3 above)
Comment (andiwand) [reply Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:29]: hmmm not sure. if it should work as a service the inheritance sounds fine. if it is just to get the logging mechanics IMO copy pasting the logging stuff would be better. but up to you. let me know how you chose and I will approve for merging @asalzburger

Comment (asalzburger) [reply Examples/Detectors/Geant4Detector/include/ActsExamples/Geant4Detector/Geant4DetectorService.hpp:29]: I take your point, will adapt accordingly tomorrow.

**Review by andiwand — APPROVED** [review]

**Comment by asalzburger** [comment]
Screenshot 2023-01-11 at 16 33 54

**Pull request merged**